include/boost/math/special_functions/prime_sieve.hpp

```diff
@@ -0,0 +1,378 @@
+// Copyright 2020 Matt Borland
+//
+// Use, modification and distribution are subject to the
+// Boost Software License, Version 1.0.
+// (See accompanying file LICENSE_1_0.txt
+// or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef BOOST_MATH_SPECIAL_FUNCTIONS_PRIME_SIEVE_HPP
+#define BOOST_MATH_SPECIAL_FUNCTIONS_PRIME_SIEVE_HPP
+
+#include <boost/math/special_functions/prime.hpp>
+#include <boost/assert.hpp>
+#include <deque>
+#include <vector>
+#include <iterator>
+#include <cmath>
+#include <thread>
+
+#if __cplusplus >= 201703
+#include <execution>
+#endif
+
+namespace boost { namespace math { namespace detail
+{
+// https://mathworld.wolfram.com/SieveofEratosthenes.html
+// https://www.cs.utexas.edu/users/misra/scannedPdf.dir/linearSieve.pdf
+template<class Z, class Container>
+void linear_sieve(Z upper_bound, Container &c)
+{
+    size_t least_divisors_size{static_cast<size_t>(upper_bound + 1)};
+    Z *least_divisors{new Z[least_divisors_size]{0}};
+
+    for (Z i{2}; i <= upper_bound; ++i)
+    {
+        if (least_divisors[i] == 0)
+        {
+            least_divisors[i] = i;
+            c.emplace_back(i);
+        }
+
+        for (size_t j{}; j < least_divisors_size; ++j)
+        {
+            if (j >= c.size())
+            {
+                break;
+            }
+
+            else if (c[j] > least_divisors[i])
+            {
+                break;
+            }
+
+            else if (i * c[j] > upper_bound)
+            {
+                break;
+            }
+
+            else
+            {
+                least_divisors[i * c[j]] = c[j];
+            }
+        }
+    }
+
+    delete[] least_divisors;
+}
+
+template<class Z, class Container>
+void prime_table(Z upper_bound, Container &c)
+{
+    Z i{2};
+    unsigned counter{};
+
+    while (i <= upper_bound && counter < 9999) // 10k elements are in the lookup table
+    {
+        c.emplace_back(i);
+        ++counter;
+        i = static_cast<Z>(boost::math::prime(counter));
+    }
+}
+
+//https://core.ac.uk/download/pdf/62440589.pdf
+template<typename Z, class PrimeContainer, class Container>
+void mask_sieve(Z lower_bound, Z upper_bound, const PrimeContainer& primes, Container &c)
+{
+    Z limit {static_cast<Z>(std::floor(std::sqrt(static_cast<double>(upper_bound)))) + 1};
+
+    size_t primes_size;
+    auto it{primes.begin()};
+    while(*it < limit && it != primes.end())
```

Comment:
This dereferences one past the end: swapping the checks over so we check for it != primes.end() first fixes that.